specifyweb/specify/tree_views.py

```diff
@@ -343,4 +351,392 @@ def perm_target(tree):
+    'COUNTRY': 200,
+    'STATE': 300,
+    'COUNTY': 400
+}
+STORAGE_RANKS = {
+    'BUILDING': 100,
+    'COLLECTION': 150,
+    'ROOM': 200,
+    'AISLE': 250,
+    'CABINET': 300,
+    'SHELF': 350,
+    'BOX': 400,
+    'RACK': 450,
+    'VIAL': 500
+}
+GEOLOGIC_TIME_PERIOD_RANKS = {
+    'ERA': 100,
+    'PERIOD': 200,
+    'EPOCH': 300,
+    'AGE': 400
+}
+LITHO_STRAT_RANKS = {
+    'SUPERGROUP': 100,
+    'GROUP': 200,
+    'FORMATION': 300,
+    'MEMBER': 400,
+    'BED': 500
+}
+
+DEFAULT_RANK_INCREMENT = 100
+TAXON_RANK_INCREMENT = 10
+GEOGRAPHY_RANK_INCREMENT = DEFAULT_RANK_INCREMENT
+STORAGE_RANK_INCREMENT = 50
+GEOLOGIC_TIME_PERIOD_INCREMENT = DEFAULT_RANK_INCREMENT
+LITHO_STRAT_INCREMENT = DEFAULT_RANK_INCREMENT
+
+# Map tree type to default tree ranks and default rank id increment
+TREE_RANKS_MAPPING = {
+    'taxon': (TAXON_RANKS, TAXON_RANK_INCREMENT),
+    'geography': (GEOGRAPHY_RANKS, GEOGRAPHY_RANK_INCREMENT),
+    'storage': (STORAGE_RANKS, STORAGE_RANK_INCREMENT),
+    'geologictimeperiod': (GEOLOGIC_TIME_PERIOD_RANKS, GEOLOGIC_TIME_PERIOD_INCREMENT),
+    'lithostrat': (LITHO_STRAT_RANKS, LITHO_STRAT_INCREMENT),
+}
+
+# Wrapper for tree editing api functions
+def tree_edit(edit_func):
+    @login_maybe_required
+    @require_POST
+    @transaction.atomic
+    @wraps(edit_func)
+    def wrapper(*args, **kwargs):
+        try:
+            edit_func(*args, **kwargs)
+            result = {'success': True}
+        except BusinessRuleException as e:
+            result = {'success': False, 'error': str(e)}
+            return HttpResponseServerError(toJson(result), content_type="application/json")
+        return HttpResponse(toJson(result), content_type="application/json")
+
+    return wrapper
+
+@openapi(schema={
+    'post': {
+        "requestBody": {
+            "required": True,
+            "description": "Add rank to a tree.",
+            "content": {
+                "application/json": {
+                    "schema": {
+                        "type": "object",
+                        "description": "The request body.",
+                        "properties": {
+                            "newRankName": {
+                                "type": "string",
+                                "description": "The name of the new rank to add."
+                            },
+                            "parentRankName": {
+                                "type": "string",
+                                "description": "The name of the parent rank to add the new rank to (use 'root' to add to the front)."
+                            },
+                            "treeName": {
+                                "type": "string",
+                                "description": "The name of tree."
+                            },
+                            "treeID": {
+                                "type": "integer",
+                                "description": "The ID of the tree (defaults to the first tree)."
```

Comment:
what do you mean defaults to the first tree? this argument should be required

Reply:
So, in just about every database I have looked at, the Geographytreedef, Geologictimeperiodtreedef, Lithostrattreedef, Storagetreedefitem, and Taxontreedef all have just one record (one tree).  So, defaulting to the first tree, when the parameter is not given, seemed to be the usual case.  Now, with using the resource API, the front-end can set this value if needed.

Reply:
I added treeDef on the front-end, so it will be passed in the tree rank resource API call.